core/services/relay/evm/types/types.go

```diff
@@ -1,57 +1,116 @@
 package types
 
 import (
+	"bytes"
 	"context"
 	"encoding/json"
-	"errors"
 	"fmt"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/lib/pq"
-
-	"gopkg.in/guregu/null.v2"
-
+	"github.com/pkg/errors"
 	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"gopkg.in/guregu/null.v2"
 
+	"github.com/smartcontractkit/chainlink-common/pkg/codec"
 	"github.com/smartcontractkit/chainlink-common/pkg/services"
 	"github.com/smartcontractkit/chainlink-common/pkg/types"
-	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
 
 	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/utils/big"
 )
 
 type ChainReaderConfig struct {
-	// ChainContractReaders key is contract name
-	ChainContractReaders map[string]ChainContractReader `json:"chainContractReaders"`
+	// Contracts key is contract name
+	Contracts map[string]ChainContractReader `json:"contracts" toml:"contracts"`
+}
+
+type CodecConfig struct {
+	// Configs key is the type's name for the codec
+	Configs map[string]ChainCodecConfig `json:"configs" toml:"configs"`
+}
+
+type ChainCodecConfig struct {
+	TypeABI         string                `json:"typeAbi" toml:"typeABI"`
+	ModifierConfigs codec.ModifiersConfig `toml:"modifierConfigs,omitempty"`
 }
 
 type ChainContractReader struct {
-	ContractABI string `json:"contractABI"`
-	// ChainReaderDefinitions key is chainAgnostic read name.
-	ChainReaderDefinitions map[string]ChainReaderDefinition `json:"chainReaderDefinitions"`
+	ContractABI string `json:"contractABI" toml:"contractABI"`
+	// key is genericName from config
+	Configs map[string]*ChainReaderDefinition `json:"configs" toml:"configs"`
+}
+
+type ChainReaderDefinition chainReaderDefinitionFields
+
+// chainReaderDefinitionFields has the fields for ChainReaderDefinition but no methods.
+// This is necessary because package json recognizes the text encoding methods used for TOML,
+// and would infinitely recurse on itself.
+type chainReaderDefinitionFields struct {
+	CacheEnabled bool `json:"cacheEnabled,omitempty"`
+	// chain specific contract method name or event type.
+	ChainSpecificName   string                `json:"chainSpecificName"`
+	ReadType            ReadType              `json:"readType,omitempty"`
+	InputModifications  codec.ModifiersConfig `json:"input_modifications,omitempty"`
+	OutputModifications codec.ModifiersConfig `json:"output_modifications,omitempty"`
```

Comment:
I forgot to suggest this in my PR:
```suggestion
	InputModifications  codec.ModifiersConfig `json:"inputModifications,omitempty"`
	OutputModifications codec.ModifiersConfig `json:"outputModifications,omitempty"`
```